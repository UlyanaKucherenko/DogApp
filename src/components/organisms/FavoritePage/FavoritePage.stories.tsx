import type { Meta, StoryObj } from '@storybook/react';

import { FavoritePage } from '.';

const meta = {
  title: 'Organisms/FavoritePage',
  component: FavoritePage,
  argTypes: {},
  parameters: {
    layout: 'fullscreen'
  }
} satisfies Meta<typeof FavoritePage>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Page: Story = {
  render: () => <FavoritePage />
};
