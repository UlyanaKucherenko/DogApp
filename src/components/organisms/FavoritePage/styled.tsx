import { Box } from '@mui/material';
import { styled } from '@mui/material/styles';

const Wrap = styled(Box)(() => ({
  paddingTop: '150px'
}));

const Container = styled(Box)(() => ({
  position: 'relative',
  maxWidth: 1200,
  width: '100%',
  margin: '0 auto',
  padding: '0 15px'
}));

const ShadowTopWrap = styled(Box)(() => ({
  position: 'absolute',
  top: 0,
  right: 0,
  zIndex: -1,
  maxWidth: '400px',

  img: {
    width: '100%'
  }
}));

const ShadowBottomWrap = styled(Box)(() => ({
  position: 'absolute',
  left: 0,
  bottom: '-24px',
  zIndex: -1,
  maxWidth: '400px',

  img: {
    width: '100%'
  }
}));

export { Wrap, Container, ShadowTopWrap, ShadowBottomWrap };
