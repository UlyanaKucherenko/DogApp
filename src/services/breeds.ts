import { api } from './api';

export const breedsApi = api.injectEndpoints({
  endpoints: build => ({
    getBreeds: build.query<Breed[], void>({
      query: () => 'breeds'
    }),
    getBreedById: build.query<BreedDetails, string>({
      query: breedId => `breeds/${breedId}`
    })
  })
});

export const { useGetBreedsQuery, useGetBreedByIdQuery } = breedsApi;

export interface Breed {
  id: string;
  name: string;
  temperament: string;
}

export interface BreedDetails extends Breed {
  weight: {
    metric: string;
  };
  height: {
    metric: string;
  };
  lifeSpan: string;
  origin: string;
}
