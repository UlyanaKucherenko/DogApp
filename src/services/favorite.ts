import { api } from './api';

export interface Favorite {
  image_id: string;
  sub_id?: string;
}

export type FavoritesResponse = {
  id: number;
  user_id: string;
  image_id: string;
  sub_id: string;
  created_at: string;
  image: {
    id: string;
    url: string;
  };
};

export const favoritesApi = api.injectEndpoints({
  endpoints: build => ({
    addFavorite: build.mutation<unknown, Favorite>({
      query: body => ({
        url: `favourites`,
        method: 'POST',
        body
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    deleteFavorite: build.mutation<unknown, { favourite_id: number }>({
      query: ({ favourite_id }) => ({
        url: `favourites/${favourite_id}`,
        method: 'DELETE'
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    getFavorites: build.query<FavoritesResponse[], void>({
      query: () => ({
        url: `favourites`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Favorites', id }) as const),
        { type: 'Favorites' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useAddFavoriteMutation,
  useDeleteFavoriteMutation,
  useGetFavoritesQuery
} = favoritesApi;
