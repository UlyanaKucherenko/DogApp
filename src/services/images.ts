import { api } from './api';

export interface Images {
  id: string;
  url: string;
  width: number;
  height: number;
  mime_type: string;
  breeds: [
    {
      id: number;
      name: string;
      weight: string;
      height: string;
      life_span: string;
      reference_image_id: string;
    }
  ];
  categories: [];
}

export interface UploadImage {
  breeds: [];
  id: string;
  url: string;
  width: number;
  height: number;
  sub_id: string;
  created_at: string;
  original_filename: string;
  breed_ids: string;
}

export type ImagesUploadResponse = {
  id: string;
  url: string;
  width: number;
  height: number;
  original_filename: string;
  pending: number;
  approved: number;
};

export interface BreedImage {
  id: string;
  url: string;
  reference_image_id: string;
}

type ImagesResponse = {
  [breedId: string]: BreedImage[];
};

export const imagesApi = api.injectEndpoints({
  endpoints: build => ({
    getBreedImages: build.query<ImagesResponse, string>({
      query: imageId => ({
        url: `images/${imageId}`
      }),
      providesTags: id => [{ type: 'Images', id }]
    }),

    getImagesId: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}&breed_id=${breed_id}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    getImages: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    uploadImage: build.mutation<Images, FormData>({
      query(formData) {
        return {
          url: `images/upload`,
          method: 'POST',
          body: formData
        };
      },
      invalidatesTags: [{ type: 'Images', id: 'LIST' }]
    }),
    uploadImagesList: build.query<
      UploadImage[],
      {
        limit?: number;
        page?: number;
      }
    >({
      query: ({ limit = 10, page = 0 }) => ({
        url: `images/?limit=${limit}&page=${page}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useGetBreedImagesQuery,
  useGetImagesIdQuery,
  useGetImagesQuery,
  useUploadImageMutation,
  useUploadImagesListQuery
} = imagesApi;
