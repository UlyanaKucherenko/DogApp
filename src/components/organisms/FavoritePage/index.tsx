import shadowBottom from '../../../assets/img/mainPage/shadow/shadow-1.svg';
import shadowTop from '../../../assets/img/mainPage/shadow/shadow-2.svg';
import { useGetFavoritesQuery } from '../../../services/favorite';
import { Loader } from '../../atoms/Loader';
import { FavoriteList } from '../FavoriteList';
import { Wrap, Container, ShadowTopWrap, ShadowBottomWrap } from './styled';

export const FavoritePage = () => {
  const { data, isLoading } = useGetFavoritesQuery();

  return (
    <Wrap>
      <ShadowTopWrap>
        <img src={shadowTop} alt="" />
      </ShadowTopWrap>
      <Container>
        {isLoading && <Loader />}
        {data && <FavoriteList favoritesData={data} />}
      </Container>
      <ShadowBottomWrap>
        <img src={shadowBottom} alt="" />
      </ShadowBottomWrap>
    </Wrap>
  );
};
