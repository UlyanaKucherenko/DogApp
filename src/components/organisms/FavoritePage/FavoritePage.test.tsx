import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import * as stories from './FavoritePage.stories';
import { store } from '../../../services/store';

describe('FavoritePage', () => {
  const { Page } = composeStories(stories);
  it('should match snapshot FavoritePage', () => {
    const { container } = render(
      <Provider store={store}>
        <Page />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
