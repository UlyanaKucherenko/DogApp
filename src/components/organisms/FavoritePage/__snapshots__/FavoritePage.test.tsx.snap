// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritePage should match snapshot FavoritePage 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-1 {
  0% {
    stroke-dasharray: 1px,200px;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -15px;
  }

  100% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -125px;
  }
}

.emotion-0 {
  padding-top: 150px;
}

.emotion-1 {
  position: absolute;
  top: 0;
  right: 0;
  z-index: -1;
  max-width: 400px;
}

.emotion-1 img {
  width: 100%;
}

.emotion-2 {
  position: relative;
  max-width: 1200px;
  width: 100%;
  margin: 0 auto;
  padding: 0 15px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 400px;
}

.emotion-4 {
  display: inline-block;
  -webkit-animation: animation-0 1.4s linear infinite;
  animation: animation-0 1.4s linear infinite;
}

.emotion-5 {
  display: block;
}

.emotion-6 {
  stroke: currentColor;
  stroke-dasharray: 80px,200px;
  stroke-dashoffset: 0;
  -webkit-animation: animation-1 1.4s ease-in-out infinite;
  animation: animation-1 1.4s ease-in-out infinite;
}

.emotion-7 {
  position: absolute;
  left: 0;
  bottom: -24px;
  z-index: -1;
  max-width: 400px;
}

.emotion-7 img {
  width: 100%;
}

<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <img
        alt=""
        src="test-file-stub"
      />
    </div>
    <div
      class="MuiBox-root emotion-2"
    >
      <div
        class="MuiBox-root emotion-3"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit emotion-4"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg emotion-5"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate emotion-6"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="MuiBox-root emotion-7"
    >
      <img
        alt=""
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
